### PR TITLE
Java: Type based summary models.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we introduce the a query for generating *type based* models for Java.
This is the Java version of https://github.com/github/codeql/pull/10238.

It is worth mentioning the following differences.
- Java does not have a `Func` or `delegate` like type, but instead relies on *functional interfaces* (ie. interfaces with a single abstract method) as these can be perceived as functions, which complicated the implementation compared to C#.
- Java has different way of providing accesspaths into collections depending on the collection type.
- As a reference example `java.util.stream.Stream` is used (an approximate equivalent to the `System.Linq.Enumerable` extension methods for C#).

### CODE REVIEW

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:7]: Something's weird here. We're importing `java.util.stream.*` such that the standard `Stream` class is in scope, but we're also defining a class with the same name, which I guess shadows the imported name?? Are we attempting to stub-implement `java.util.stream.Stream` or import it?

File: java/ql/test/utils/model-generator/typebasedflow/p/TypeBasedComplex.java
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/p/TypeBasedComplex.java:19]: This type signature doesn't refer to `T`. Was that intentional or was the method perhaps meant to be static?

Comment (aschackmull) [reply java/ql/test/utils/model-generator/typebasedflow/p/TypeBasedComplex.java:19]: If it's meant as-is, then it deserves an explanatory comment.

File: java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected:1]: The naming of the synthetic fields could be a bit more precise. In particular they ought to refer to the type (when they aren't derived from a super-class/-interface), or reuse the field that the supertype indicates. I.e. a class `MyClass<T>` can use `SyntheticField[package.MyClass.T]`, a class `MySubclass<T2> extends MyClass<T2>` should use the same field as the superclass `SyntheticField[package.MyClass.T]`, and an extension of known collection-types should reuse those, such that `MyMultiMap<K,V> implements Map<K,List<V>>` uses `MapKey` and `MapValue.Element`. (The test `TypeBasedCollection<T> extends ArrayList<T>` already checks that to some degree).
Note that this will also ensure that methods on `A` and `B` will be compatible when we have `class A<T1,T2> extends B<T2>`, whereas the current approach fails to do so (equating the wrong fields and misses the equality that ought to be present).

Comment (aschackmull) [reply java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected:1]: Given that a class can implement multiple interfaces, then we could have a conflict, so in that case the preference order should probably be something like "known collection-like super-interface" > "super-class" > "super-interface".

Comment (aschackmull) [reply java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected:1]: Also, could we make this into some sort of inline test expectation? We don't necessarily need to use the inline expectation library - a more ad-hoc approach could be fine: I'm thinking perhaps collect all comments of the form `// MaD = ...` and then compare the query output with the set of those strings. We don't necessarily need to make use of the comment locations, which means that we could just write several expectation lines above the method that generates them.

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (aschackmull) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:7]: I wonder then why the generated models use `Element`? I mean they certainly ought to for `java.util.stream.Stream` but this stub class isn't that. If the stub class was implementing, say, `Iterable<T>` then it would make more sense, as that ought to be enough to make it use `Element` as opposed to generating a synthetic field.

Comment (aschackmull) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:7]: When we're stubbing like this, then I think we should use explicit imports from `java.util.stream` such that we aren't shadowing anything. That would make it more clear that the references to `Stream` are in fact this stub class.

File: java/ql/test/utils/model-generator/typebasedflow/p/TypeBasedComplex.java
Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/p/TypeBasedComplex.java:19]: Yes, that as intentional. The test is there to check that no summary that accesses the *implicit* field corresponding to the field of type `T` will be generated (even for instance methods).

File: java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected
Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected:1]: > The naming of the synthetic fields could be a bit more precise. In particular they ought to refer to the type (when they aren't derived from a super-class/-interface), or reuse the field that the supertype indicates. I.e. a class `MyClass<T>` can use `SyntheticField[package.MyClass.T]`, a class `MySubclass<T2> extends MyClass<T2>` should use the same field as the superclass `SyntheticField[package.MyClass.T]`, and an extension of known collection-types should reuse those, such that `MyMultiMap<K,V> implements Map<K,List<V>>` uses `MapKey` and `MapValue.Element`. (The test `TypeBasedCollection<T> extends ArrayList<T>` already checks that to some degree). Note that this will also ensure that methods on `A` and `B` will be compatible when we have `class A<T1,T2> extends B<T2>`, whereas the current approach fails to do so (equating the wrong fields and misses the equality that ought to be present).

You are absolutely right that the implementation could be improved and that the precision for more complicated inheritance is not ideal. The same is the case for C#, which also lacks exactly what you are describing here.
Do you prefer that we get this sorted as a first attempt before merging this?

Comment (aschackmull) [reply java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected:1]: > Do you prefer that we get this sorted out as a first attempt before merging this?

I'll leave that to your preference. But I would like a refactor of the test, such that the models could be seen directly in the Java source next to the method.

Comment (aschackmull) [reply java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected:1]: Something like
```
string expects() {
  exists(JavaDoc doc |
    doc.getChild(0).toString().regexpCature(" *MaD *= *\\(.*\\)", 1) = result
  )
}

...
```

Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected]: Yes, this is a great idea. I have included in it in this PR for the type based summaries, but maybe we should consider to do the same thing for both Java and C# for all model generator like tests.

File: java/ql/test/utils/model-generator/typebasedflow/p/TypeBasedComplex.java
Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/p/TypeBasedComplex.java:27]: Comment added.

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:13]: (1) As far as can tell from the documentation of the `Stream` (java) interface and `ContainerType` (QL) implementation, then `Stream` will not be identified as a `ContainerType` (note that `Stream<T>` does not implement `Iterable<T>`). The implementation on this PR will also perceive classes or interfaces having any method that returns an `Iterator` as being a container like type (but this purely has effect the type based model generation implementation. 
(2) Imports has been changed to avoid confusion.

File: java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.ql
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.ql:13]: Simplify:
```suggestion
    not flow = expects() and
```

File: java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.ql
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.ql:21]:
```suggestion
    not e = flows() and
```

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:70]: Why is this test case excluded?

Comment (aschackmull) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:70]: If it doesn't work, then we might as well keep it with a comment about why and what doesn't work.

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:103]: All four of these models look correct to me?

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:158]: That's fine, but no reason to comment the test case.

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:179]: Yeah, this one is spurious but mostly harmless. Maybe it would be nice if the regex for matching `MaD=` would allow a `SPURIOUS` prefix so we could write
```suggestion
    // SPURIOUS MaD=p;Stream;true;reduce;(BinaryOperator);;Argument[0].ReturnValue;Argument[-1].Element;value;generated
```

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (aschackmull) [inline java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:174]: s/a couple of/one/

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:17]:
```suggestion
      return.getSourceDeclaration().hasQualifiedName("java.util", "Iterator") and
```

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:15]:
```suggestion
    exists(Method m, RefType return, GenericType t, Type et, int position | m.getDeclaringType() = t |
```

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:48]: No? In Java they aren't. Or is this something that you assume for some heuristic purpose?

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:120]: This will need some refinement at some point, but it's probably good enough as a starting point.

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:11]: There are some implicit assumptions throughout this file that an `IterableType` has exactly one type variable. This is probably fine as a starting point, as that's likely to generally be the case, but might deserve a comment.

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:165]:
```suggestion
   * Note that functions are often contravariant in their parameter types.
```

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:176]:
```suggestion
   * Note that functions are often covariant in their return type.
```

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:194]: Simplify:
```suggestion
  callable.getParameterType(position) = f
```

**Review by aschackmull — COMMENTED** [review]
Mostly minor comments, otherwise LGTM!

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (michaelnebel) [reply java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:48]: Ok, that claim is to strong.
At least the idea is that, if we have "relevant" type variable that serves as an upper bound for a given type, then we use that to eg. decide if the method reads from an implicit field or the implicit collection.
I will delete the comment.

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (michaelnebel) [reply java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:120]: Yes, we do not dig into the inheritance hierarchy at the moment, but that needs to be improved.

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (michaelnebel) [reply java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:11]: An IterableType can be parameterized over multiple type variables. As an example
```java
public class Stream2<T1, T2> {

    public Iterator<T1> iterator() {
        return null;
    }

    public T2 getField() {
        return null;
    }

    public T1 getElement() {
        return null;
    }
}
```
produce the models
```
p;Stream2;true;getElement;();;Argument[-1].Element;ReturnValue;value;generated
p;Stream2;true;getField;();;Argument[-1].SyntheticField[ArgType1];ReturnValue;value;generated p;Stream2;true;iterator;();;Argument[-1].Element;ReturnValue.Element;value;generated
```

I am not sure where you discovered the implicit assumptions.
If you have seen something, then it might be worth narrowing the InterableType definition, such we only accept types that are parameterized over a single type variable and where there exists an iterator method over that exact type.

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:70]: That is an oversight on my part - it does work and I will comment it back in.

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:103]: Yes, you are right! I will remove the comment (I had overlooked that the "seed" is the first element of the sequence, which is why I thought that Argument[0] -> ReturnValue.Element was incorrect - but that is not the case).

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:158]: I will comment it in.

File: java/ql/test/utils/model-generator/typebasedflow/p/Stream.java
Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:179]: Also: Argument[0].ReturnValue -> Argument[0].Parameter[1] is spurious. In the documentation it is stated that only required is that the operator is associative.

Comment (michaelnebel) [reply java/ql/test/utils/model-generator/typebasedflow/p/Stream.java:179]: Hmm - looks like that one is included in the manual models.

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [reply java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll]: Consider this (probably an obscure corner-case, but still):
```
public class Stream2<T1, T2> {
    public Iterator<T1> iterator1() {
        return null;
    }

    public Iterator<T2> iterator2() {
        return null;
    }

    public T1 getElement1() {
        return null;
    }

    public T2 getElement2() {
        return null;
    }
}
```

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (aschackmull) [reply java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:50]: Not yet deleted.

Comment (michaelnebel) [reply java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll:50]: haha - no - you are right! I will try once more :-)

File: java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll
Comment (michaelnebel) [reply java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll]: You are right. In that case, we probably shouldn't think of `Stream2` as being an interable type.
It is an easy fix.

**Review by aschackmull — APPROVED** [review]

**Review by Saidyyh — COMMENTED** [review]
2011685e89656c624fb2fe8e4ae3e16715727186

**Pull request merged**
